File: 20230901_Friday/20230901.java
// Leetcode: 338. Counting Bits.
// https://leetcode.com/problems/counting-bits/
class Solution {
  // Copied from:
  // https://leetcode.com/problems/counting-bits/discuss/1808002/A-very-very-EASY-to-go-EXPLANATION
    public int[] countBits(int n) {
        int res[] = new int[n + 1]; // create one result array, 
        // & our array size is n + 1 because we have to cover 0 as well
        
        for(int i = 0; i <= n; i++){ // run a loop to store all the values we get from a solve function
            res[i] = solve(i);
        }
        return res;
    }
    public int solve(int n){
        // base condition
        if(n == 0) return 0;
        if(n == 1) return 1;
        
        if(n % 2 == 0) return solve(n / 2); // handling even case
        else return 1 + solve(n / 2); // handling odd case
    }
}

File: 20230901_Friday/20230901.rb
# Leetcode: 338. Counting Bits.
# https://leetcode.com/problems/counting-bits/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 104 ms, faster than 90.36% of Ruby online submissions for Counting
# Bits.
# Memory Usage: 216 MB, less than 68.67% of Ruby online submissions for Counting
# Bits.
# 2023.09.01 Daily Challenge.
# @param {Integer} n
# @return {Integer[]}
def count_bits(n)
  # Based on:
  # https://leetcode.com/problems/counting-bits/discuss/1808002/A-very-very-EASY-to-go-EXPLANATION
  return [0] if 0 == n
  return [0,1] if 1 == n
  ans = Array.new(n+1, 0)
  ans[0] = 0
  ans[1] = 1
  (2..n).each do |i|
    if i.even?
      ans[i] = ans[i/2]
    else
      ans[i] = 1 + ans[i/2]
    end
  end
  return ans
end

